modules/skunk/src/test/scala/mutation/MutationData.scala

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) 2016-2020 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package mutation
+
+import cats.effect.{ Bracket, Resource, Sync }
+import cats.syntax.all._
+import edu.gemini.grackle._
+import edu.gemini.grackle.QueryCompiler._
+import edu.gemini.grackle.Query._
+import edu.gemini.grackle.Value._
+import edu.gemini.grackle.Predicate._
+import edu.gemini.grackle.skunk._
+import _root_.skunk.Session
+import _root_.skunk.implicits._
+import _root_.skunk.codec.all._
+import cats.data.IorT
+
+trait MutationSchema[F[_]] extends SkunkMapping[F] {
+
+  class TableDef(name: String) {
+    def col(colName: String, codec: Codec[_]): ColumnRef =
+      ColumnRef(name, colName, codec)
+  }
+
+  object country extends TableDef("country") {
+    val code = col("code", bpchar(3))
+    val name = col("name", text)
+  }
+
+  object city extends TableDef("city") {
+    val id          = col("id", int4)
+    val countrycode = col("countrycode", bpchar(3))
+    val name        = col("name", text)
+    val population  = col("population", int4)
+  }
+
+}
+
+trait MutationMapping[F[_]] extends MutationSchema[F] {
+
+  implicit def ev: Bracket[F, Throwable]
+
+  val schema =
+    Schema(
+      """
+        type Query {
+          city(id: Int!): City
+        }
+        type Mutation {
+          updatePopulation(id: Int!, population: Int!): City
+        }
+        type City {
+          name: String!
+          country: Country!
+          population: Int!
+        }
+        type Country {
+          name: String!
+          cities: [City!]!
+        }
+      """
+    ).right.get
+
+  val QueryType    = schema.ref("Query")
+  val MutationType = schema.ref("Mutation")
+  val CountryType  = schema.ref("Country")
+  val CityType     = schema.ref("City")
+
+  val typeMappings =
+    List(
+      ObjectMapping(
+        tpe = QueryType,
+        fieldMappings = List(
+          SqlRoot("city"),
+        ),
+      ),
+      ObjectMapping(
+        tpe = MutationType,
+        fieldMappings = List(
+          SqlRoot("updatePopulation", mutation = Mutation.unit { (_, e) =>
+            (e.get[Int]("id"), e.get[Int]("population")).tupled match {
+              case None =>
+                QueryInterpreter.mkErrorResult(s"Implementation error, expected id and population in $e.").pure[F]
+              case Some((id, pop)) =>
+                pool.use { s =>
+                  s.prepare(sql"update city set population=$int4 where id=$int4".command).use { ps =>
+                    IorT.right(ps.execute(pop ~ id).void).value // awkward
+                  }
+                }
+              }
+          }),
+        ),
+      ),
```

Comment:
The combined ergonomics of `Env` and `Result` and `F` make this a little awkward.

Reply:
Agreed, but this seems like a good starting point at least.